67_RayQueryGeometry/main.cpp

```diff
@@ -0,0 +1,1043 @@
+						if (!buildSizes)
+							return logFail("Failed to get BLAS build sizes");
+					}
+
+					scratchSizes[i] = buildSizes.buildScratchSize;
+					totalScratchSize += buildSizes.buildScratchSize;
+
+					{
+						IGPUBuffer::SCreationParams params;
+						params.usage = bitflag(IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT) | IGPUBuffer::EUF_ACCELERATION_STRUCTURE_STORAGE_BIT;
+						params.size = buildSizes.accelerationStructureSize;
+						smart_refctd_ptr<IGPUBuffer> asBuffer = createBuffer(params);
+
+						IGPUBottomLevelAccelerationStructure::SCreationParams blasParams;
+						blasParams.bufferRange.buffer = asBuffer;
+						blasParams.bufferRange.offset = 0u;
+						blasParams.bufferRange.size = buildSizes.accelerationStructureSize;
+						blasParams.flags = IGPUBottomLevelAccelerationStructure::SCreationParams::FLAGS::NONE;
+						gpuBlas[i] = m_device->createBottomLevelAccelerationStructure(std::move(blasParams));
+						if (!gpuBlas[i])
+							return logFail("Could not create BLAS");
+					}
+				}
+
+				auto cmdbufBlas = getSingleUseCommandBufferAndBegin(pool);
+				cmdbufBlas->beginDebugMarker("Build BLAS");
+
+				cmdbufBlas->resetQueryPool(queryPool.get(), 0, objectsGpu.size());
+
+				smart_refctd_ptr<IGPUBuffer> scratchBuffer;
+				{
+					IGPUBuffer::SCreationParams params;
+					params.usage = bitflag(IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT) | IGPUBuffer::EUF_STORAGE_BUFFER_BIT;
+					params.size = totalScratchSize;
+					scratchBuffer = createBuffer(params);
+				}
+
+				uint32_t queryCount = 0;
+				IGPUBottomLevelAccelerationStructure::BuildRangeInfo buildRangeInfos[OT_COUNT];
+				IGPUBottomLevelAccelerationStructure::BuildRangeInfo* pRangeInfos[OT_COUNT];
+				for (uint32_t i = 0; i < objectsGpu.size(); i++)
+				{
+					blasBuildInfos[i].dstAS = gpuBlas[i].get();
+					blasBuildInfos[i].scratch.buffer = scratchBuffer;
+					blasBuildInfos[i].scratch.offset = (i == 0) ? 0u : blasBuildInfos[i - 1].scratch.offset + scratchSizes[i - 1];
+
+					buildRangeInfos[i].primitiveCount = primitiveCounts[i];
+					buildRangeInfos[i].primitiveByteOffset = 0u;
+					buildRangeInfos[i].firstVertex = 0u;
+					buildRangeInfos[i].transformByteOffset = 0u;
+
+					pRangeInfos[i] = &buildRangeInfos[i];
+				}
+
+				if (!cmdbufBlas->buildAccelerationStructures({ blasBuildInfos, OT_COUNT }, pRangeInfos))
+					return logFail("Failed to build BLAS");
+
+				{
+					SMemoryBarrier memBarrier;
+					memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::ACCELERATION_STRUCTURE_BUILD_BIT;
+					memBarrier.srcAccessMask = ACCESS_FLAGS::ACCELERATION_STRUCTURE_WRITE_BIT;
+					memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::ACCELERATION_STRUCTURE_BUILD_BIT;
+					memBarrier.dstAccessMask = ACCESS_FLAGS::ACCELERATION_STRUCTURE_READ_BIT;
+					cmdbufBlas->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, { .memBarriers = {&memBarrier, 1} });
+				}
```

Comment:
id you submit and wait for the command buffer, you don't need to do a pipeline barrier, the semaphore already does it